sound/soc/sof/pcm.c

```diff
@@ -52,6 +52,39 @@ static int sof_pcm_dsp_params(struct snd_sof_pcm *spcm, struct snd_pcm_substream
 	return ret;
 }
 
+/*
```

Comment:
please fix your commit message. Simple sentences with a period at the end. Repeated commas are totally misleading for the reader, it's not clear if you are still in the same logical proposition or moved on.

ASoC: SOF: add period_elapsed work to fix race condition in interrupt context

The IPC implementation in SOF requires sending IPCs serially: we should not send a new IPC command to the firmware before we get an ACK (or time out) from firmware, and the IRQ processing is complete.

snd_pcm_period_elapsed() can be called in interrupt context before IRQ_HANDLED is returned. When the PCM is done draining, a STOP IPC will then be send, which breaks the expectation that IPCs are handled serially and leads to IPC timeouts.

This patch adds a workqueue to defer the call to snd_pcm_elapsed() after the IRQ is handled.